rcljava/src/main/java/org/ros2/rcljava/action/ActionServerImpl.java

```diff
@@ -509,8 +614,53 @@ public void execute() {
     }
 
     if (this.isResultRequestReady()) {
-      // executeResultRequest(rmwRequestId, requestMessage, responseMessage);
-      // TODO
+      Class<? extends ResultRequestDefinition> requestType = this.actionTypeInstance.getGetResultRequestType();
+
+      ResultRequestDefinition<T> requestMessage = null;
+      try {
+        requestMessage = requestType.newInstance();
+      } catch (Exception ex) {
+        throw new IllegalArgumentException("Failed to instantiate action result request", ex);
+      }
+
+      if (requestMessage != null) {
+        long requestFromJavaConverterHandle = requestMessage.getFromJavaConverterInstance();
+        long requestToJavaConverterHandle = requestMessage.getToJavaConverterInstance();
+        long requestDestructorHandle = requestMessage.getDestructorInstance();
+
+        RMWRequestId rmwRequestId =
+          nativeTakeResultRequest(
+            this.handle,
+            requestFromJavaConverterHandle, requestToJavaConverterHandle, requestDestructorHandle,
+            requestMessage);
+
+        List<Byte> goalUuid = requestMessage.getGoalUuid();
+        boolean goalExists = this.goalExists(this.createGoalInfo(goalUuid));
+
+        ResultResponseDefinition<T> resultResponse = null;
+        if (!goalExists) {
+          // Goal does not exists
+          resultResponse = this.createResultResponse();
+          resultResponse.setGoalStatus(action_msgs.msg.GoalStatus.STATUS_UNKNOWN);
+        } else {
+          // Goal exists, check if a result is already available
+          resultResponse = this.goalResults.get(goalUuid);
+        }
+
+        if (null == resultResponse) {
+          List<RMWRequestId> requestIds = null;
+          if (rmwRequestId != null) {
+            requestIds = this.goalRequests.get(goalUuid);
+            if (requestIds == null) {
+              requestIds = new ArrayList();
+              this.goalRequests.put(goalUuid, requestIds);
+            }
+          }
+          requestIds.add(rmwRequestId);
+        } else {
+          this.sendResultResponse(rmwRequestId, resultResponse);
```

Comment:
Is it possible for `rmwRequestId` to be `null`? If so, should we guard against it?

Reply:
Yes, fixed in https://github.com/osrf/ros2_java/pull/65/commits/b6f247a03c2294e93c9992440869e16c3a247921.